contracts/upgradeable_contracts/BridgeValidators.sol

```diff
@@ -9,89 +9,167 @@ import "../upgradeability/EternalStorage.sol";
-        setValidator(_validator, false);
+        address validatorsNext = getNextValidator(_validator);
+        address index = F_ADDR;
+        address next = getNextValidator(index);
+
+        // find the element in the list pointing to _validator
+        while (next != _validator) {
+            index = next;
+            next = getNextValidator(index);
+        }
+
+        setValidator(index, validatorsNext);
+        deleteItemFromAddressStorage("validatorsList", _validator);
+        deleteItemFromAddressStorage("validatorsRewards", _validator);
         setValidatorCount(validatorCount().sub(1));
+
         emit ValidatorRemoved(_validator);
     }
 
-    function setRequiredSignatures(uint256 _requiredSignatures) external onlyOwner {
+    function setRequiredSignatures(uint256 _requiredSignatures)
+        external
+        onlyOwner
+    {
         require(validatorCount() >= _requiredSignatures);
         require(_requiredSignatures != 0);
         uintStorage[keccak256(abi.encodePacked("requiredSignatures"))] = _requiredSignatures;
         emit RequiredSignaturesChanged(_requiredSignatures);
     }
 
-    function getBridgeValidatorsInterfacesVersion() public pure returns(uint64 major, uint64 minor, uint64 patch) {
+    function getBridgeValidatorsInterfacesVersion()
+        public
+        pure
+        returns (uint64 major, uint64 minor, uint64 patch)
+    {
         return (2, 0, 0);
     }
 
-    function requiredSignatures() public view returns(uint256) {
+    function requiredSignatures() public view returns (uint256) {
         return uintStorage[keccak256(abi.encodePacked("requiredSignatures"))];
     }
 
-    function validatorCount() public view returns(uint256) {
+    function validatorCount() public view returns (uint256) {
         return uintStorage[keccak256(abi.encodePacked("validatorCount"))];
     }
 
-    function validators(address _validator) public view returns(bool) {
-        return boolStorage[keccak256(abi.encodePacked("validators", _validator))];
-    }
-
-    function isValidator(address _validator) public view returns(bool) {
-        return validators(_validator) == true;
+    function isValidator(address _validator) public view returns (bool) {
+        return _validator != F_ADDR && getNextValidator(_validator) != address(0);
     }
 
-    function isInitialized() public view returns(bool) {
+    function isInitialized() public view returns (bool) {
         return boolStorage[keccak256(abi.encodePacked("isInitialized"))];
     }
 
-    function deployedAtBlock() public view returns(uint256) {
+    function deployedAtBlock() public view returns (uint256) {
         return uintStorage[keccak256("deployedAtBlock")];
     }
 
+    function getValidatorRewardAddress(address _validator) public view returns (address) {
+        return addressStorage[keccak256(abi.encodePacked("validatorsRewards", _validator))];
+    }
+
+    function getNextValidator(address _address) public view returns (address) {
```

Comment:
Should we revert if the corresponding element points to `adress(0)`?

Reply:
Right now there is a use case if element points to `address(0)`, it help us know if the address is a validator.
https://github.com/poanetwork/poa-bridge-contracts/blob/4e9db7b626ef73a847b019247865d9f0ae1d53d4/contracts/upgradeable_contracts/BaseBridgeValidators.sol#L41-L43

I think it's a good idea to let this method just get the value, and call `revert()` on the method that called `getNextValidator` if the scenario of the method requires that. what do you think?

Reply:
OK. Then let's go through all places where `getNextValidator` is invoked and check whether the scenario when it returns `address(0)` is valid or not. If it is not valid, we will `revert`.

Reply:
Done. Please check https://github.com/poanetwork/poa-bridge-contracts/pull/130/commits/212d7964122cd26838dfeb74f4dc584bb89796b8